espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ComposeNodeFinder.kt

```diff
@@ -57,3 +64,28 @@ fun getSemanticsNode(elementId: String): SemanticsNode =
     } catch (e: AssertionError) {
         throw StaleElementException(elementId)
     }
+
+private fun hasXpath(locator: Locator): SemanticsMatcher {
+    val parentNode = locator.elementId?.let { getSemanticsNode(it) }
+    val xpathSelector = locator.value!!
+    val sourceDocument = SourceDocument(parentNode, fetchIncludedAttributes(xpathSelector))
+    val expr = try {
+        XPATH.compile(xpathSelector)
+    } catch (xe: XPathExpressionException) {
+        throw XPathLookupException(xpathSelector, xe.message!!)
+    }
+    return SemanticsMatcher(
+        "Matches Xpath $xpathSelector"
+    ) {
+        val nodeIndices = sourceDocument.RESOURCES_GUARD.withPermit({
```

Comment:
I don't think we should use the same resources guard here. The matcher should have a separate one

Reply:
perhaps we should just hide this internal implementation details in SourceDocument class and only pass the matcher function there as an argument

Reply:
I am not clear what you mean, can you help with an example?

Reply:
Do you mean something like this in SourceDocument
```
fun <R> hasXpath(xpathSelector: String, block: (List<Int>) -> R): R {
        val expr = try {
            XPATH.compile(xpathSelector)
        } catch (xe: XPathExpressionException) {
            throw XPathLookupException(xpathSelector, xe.message!!)
        }
        return block(nodeIndices(expr))
    }

    private fun nodeIndices(expr: XPathExpression): List<Int> {
        val nodeIndices = RESOURCES_GUARD.withPermit({
            toStream().use { xmlStream ->
                val list =
                    expr.evaluate(InputSource(xmlStream), XPathConstants.NODESET) as NodeList
                (0 until list.length).map { index ->
                    list.item(index).attributes.getNamedItem(ViewAttributesEnum.RESOURCE_ID.toString()).nodeValue.toInt()
                }
            }
        }, { performCleanup() })
        return nodeIndices
    }
```
And we invoke it from ComposeNodeFinder like below
```
private fun hasXpath(locator: Locator): SemanticsMatcher {
    return SourceDocument(
        locator.elementId?.let { getSemanticsNode(it) },
        fetchIncludedAttributes(locator.value!!)
    ).hasXpath(locator.value!!) { nodeIndices: List<Int> ->
        SemanticsMatcher("Matches Xpath ${locator.value}") { nodeIndices.contains(it.id) }
    }
}
```

Reply:
almost. hasXpath could stay where it is now and only 

```
val nodeIndices = RESOURCES_GUARD.withPermit({
            toStream().use { xmlStream ->
                val list =
                    expr.evaluate(InputSource(xmlStream), XPathConstants.NODESET) as NodeList
                (0 until list.length).map { index ->
                    list.item(index).attributes.getNamedItem(ViewAttributesEnum.RESOURCE_ID.toString()).nodeValue.toInt()
                }
            }
        }, { performCleanup() })
```

part, which would return matched node indexes and accepts the xpath expression would belong to the SourceDocument

Reply:
let me know if you are still unsure on how to design it. I could add the patch directly to this PR

Reply:
Do you mean something like this?
```
# SourceDocuemt
fun matchingNodesIndices(xpathSelector: String): List<Int> {
    val expr = try {
        XPATH.compile(xpathSelector)
    } catch (xe: XPathExpressionException) {
        throw XPathLookupException(xpathSelector, xe.message!!)
    }
    return RESOURCES_GUARD.withPermit({
        toStream().use { xmlStream ->
            val list =
                expr.evaluate(InputSource(xmlStream), XPathConstants.NODESET) as NodeList
            (0 until list.length).map { index ->
                list.item(index).attributes.getNamedItem(ViewAttributesEnum.RESOURCE_ID.toString()).nodeValue.toInt()
            }
        }
    }, { performCleanup() })
}

# ComposeNodeFinder.kt
private fun hasXpath(locator: Locator): SemanticsMatcher {
    val sourceDocument = SourceDocument(
        locator.elementId?.let { getSemanticsNode(it) },
        fetchIncludedAttributes(locator.value!!)
    )
    return SemanticsMatcher("Matches Xpath ${locator.value}") {
        sourceDocument.matchingNodesIndices(
            locator.value!!
        ).contains(it.id)
    }
}
```

Reply:
yes, this makes sense. Also consider unifying `getMatchingNodeIds` method, so it could also be reused for espresso nodes lookup and to avoid extra code duplication if possible